[DOCS] Update Docker getting started for the Stack (#1888)
elastic/stack-docs: docs/en/getting-started/get-started-docker.asciidoc

token, which is valid for 30 minutes. This token automatically applies the
security settings from your {es} cluster, authenticates to {es} with the
`kibana_system` user, and writes the security configuration to `kibana.yml`.

[discrete]
==== Prerequisites
. Install the appropriate https://docs.docker.com/get-docker/[Docker application]
for your operating system.
+
[NOTE]
====
Make sure that Docker is allotted at least 4GiB of memory. In Docker Desktop,
you configure resource usage on the Advanced tab in Preference (macOS) or
Settings (Windows).
====

. Pull the {es} Docker image.
+
--
ifeval::["{release-state}"=="unreleased"]

WARNING: Version {version} of {es} has not yet been released, so no
Docker image is currently available for this version.

endif::[]

ifeval::["{release-state}"!="unreleased"]

[source,sh,subs="attributes"]
----
docker pull docker.elastic.co/elasticsearch/elasticsearch:{version}
----

endif::[]
--

. Pull the {kib} Docker image.
+
--
ifeval::["{release-state}"=="unreleased"]

WARNING: Version {version} of {es} has not yet been released, so no
Docker image is currently available for this version.

endif::[]

> ifeval::["{release-state}"!="unreleased"]

Reviewer: Should we show the command anyway, like we do for Kibana in the following section?

[source,sh,subs="attributes"]
----
docker pull docker.elastic.co/kibana/kibana:{version}
----

endif::[]
--

[discrete]
==== Start a single-node {es} cluster and enroll {kib}

The following command starts a single-node {es} cluster for development or
testing.

. Start {es} in Docker. A password is generated for the `elastic` user and
output to the terminal, plus an enrollment token for enrolling {kib}.
+
--
ifeval::["{release-state}"!="unreleased"]
[source,sh,subs="attributes"]
----

> docker run --name es-node01 -p 9200:9200 -p 9300:9300 -it docker.elastic.co/elasticsearch/elasticsearch:{version}

Reviewer: Do we really need to expose port 9300 to the world since we are creating a single-node cluster?

Author: This is the approach we implemented in the docs for [installing Kibana on Docker](https://www.elastic.co/guide/en/kibana/8.0/docker.html#run-kibana-on-docker-for-dev). If we don't want to expose port `9300`, then I'll update it in both places.

Reviewer: Thanks

----

endif::[]
--
+
TIP: You might need to scroll back a bit in the terminal to view the password
and enrollment token.

. Copy the generated password and enrollment token and save them in a secure
location. These values are shown only when you start {es} for the first time.
+
[NOTE]
====
If you need to reset the password for the `elastic` user or other
built-in users, run the {ref}/reset-password.html[`elasticsearch-reset-password`]
tool. To generate new enrollment tokens for {kib} or {es} nodes, run the
{ref}/create-enrollment-token.html[`elasticsearch-create-enrollment-token`] tool.
These tools are available in the {es} `bin` directory of the Docker container.
For example:

[source,sh]
----

> docker exec -it es-node01 /usr/share/elasticsearch/bin/elasticsearch-reset-password

Reviewer: ```suggestion
docker exec -it es-node01 /usr/share/elasticsearch/bin/elasticsearch-reset-password -u elastic
```

Author: Thanks for pointing out this omission -- I'll update the command with the `-u` option across all of the Docker documentation.

----
====

. Copy the `http_ca.crt` security certificate from your Docker container to
your local machine.
+
[source,sh]
----

> docker cp es-node01:/usr/share/elasticsearch/config/tls_auto_config_*/http_ca.crt .

Reviewer: This command is not working, as `*` cannot be used for path expansion from outside docker. There are a few changes that we can do to make it simpler, otherwise we should list folders first or find another way.

Author: We could just list the command as `tls_auto_config_<timestamp>` and indicate that users need to modify that portion of their command to match the filename on disk. Open to alternatives if you have preferred options.

----

. Open a new terminal and verify that you can connect to your {es}
cluster by making an authenticated call, using the `http_ca.crt` file that you
copied from your Docker container. Enter the password for the `elastic` user
when prompted.
+
[source,sh]
----
curl --cacert http_ca.crt -u elastic https://localhost:9200
----
// NOTCONSOLE

> . In a new terminal session, start {kib} and connect it to your {es} container:

Reviewer: Should we add the warning in case the version is not released yet?

+
[source,sh,subs="attributes"]
----

> docker run --name kib-01 --net elastic -p 5601:5601 docker.elastic.co/kibana/kibana:{version}

Reviewer: This command fails because the `elastic` network doesn't exist.

Author: I'll add `--net elastic` to the command for starting Elasticsearch on line 106.

Reviewer: I'm not sure if this solves the problem, IIRC the network must be created explicitly.
Can we just use the default network instead?

> docker run --name kib-01 --net elastic -p 5601:5601 docker.elastic.co/kibana/kibana:{version}

Reviewer: Container name should be consistent across Elasticsearch and Kibana, e.g. `kibana01` here or `es-01` there.

Author: I'll change to `es-node01` throughout (and subsequently `es-node02` and `es-node01`) and `kib01`.

----
+
When you start {kib}, a unique link is output to your terminal.

. To access {kib}, click the generated link in your terminal.

  .. In your browser, paste the enrollment token that you copied when starting
  {es} and click the button to connect your {kib} instance with {es}.

  .. Log in to {kib} as the `elastic` user with the password that was generated

>   when you started {es}.

Reviewer: We could add clean-up  instuctions to stop and remove containers after use.

Author: ++ I'll add a section for removing the containers.

[[get-started-stack-docker]]
== Running the {stack} ("ELK") on Docker
++++
<titleabbrev>Running the {stack} on Docker</titleabbrev>
++++

The Elastic Docker registry contains Docker images for all the products in
the {stack}: https://www.docker.elastic.co/.

You can <<run-docker-secure,start the {stack} in Docker>> with security enabled
and configured by default. This option is great for quickly getting started with
{es} and {kib} for testing and development.

You can also <<get-started-docker-tls,start the {stack} with Docker Compose>> to
create a secured, multi-node cluster with a connected {kib} instance. This

> option requires more manual steps, but results in a more robust cluster.

Reviewer 2: Should we qualify what we mean with "robust cluster" here ? It reads slightly generic

Author: Yeah, I think that we can expand a bit more without getting too in depth. How about:
>This option requires more manual steps, but results in a more resilient cluster with greater capacity and reliability.

[[run-docker-secure]]
[discrete]
=== Start {es} and {kib} in Docker
Starting in {es} 8.0, security is enabled by default. With security enabled,
{stack} {security-features} require TLS encryption for the transport networking

> layer, or your cluster will fail to start.

Reviewer 2: The TLS comment makes me think as a reader that I need to so something extra because security is enabled. I think this wording made sense in 7.x but maybe not so now ? It is true, but TLS will be enable by default if users follow what we write here so maybe we shouldn't point it out in that manner in the first paragraph ?

Author: Good point. I think that we can delete the second sentence and just say:
>Starting in {es} 8.0, security is enabled and configured by default.

If you're starting a single-node {es} cluster in a Docker container, security
will be automatically enabled and configured for you. When you start {es} for
the first time, the following security configuration occurs automatically:

* {ref}/configuring-stack-security.html#stack-security-certificates[Certificates and keys] are generated
for the transport and HTTP layers.
* The Transport Layer Security (TLS) configuration settings are written to
`elasticsearch.yml`.
* A password is generated for the `elastic` user.

> * An enrollment token is generated for {kib}.

Reviewer 2: ```suggestion
* An enrollment token is generated for {kib}.
* An enrollment token is generated for other {es} nodes.
```

You can then {kibana-ref}/docker.html[start {kib}] and enter the enrollment
token, which is valid for 30 minutes. This token automatically applies the
security settings from your {es} cluster, authenticates to {es} with the

> `kibana_system` user, and writes the security configuration to `kibana.yml`.

Reviewer 2: ```suggestion
`kibana` service account, and writes the security configuration to `kibana.yml`.
```

[discrete]
==== Prerequisites
. Install the appropriate https://docs.docker.com/get-docker/[Docker application]
for your operating system.
+
[NOTE]
====

> Make sure that Docker is allotted at least 4GiB of memory. In Docker Desktop,

Reviewer 2: Is that still valid ? I run my docker containers with less with no issue

Author: It's definitely valid for running a multi-node cluster with Docker Compose. With less memory than this, one node in my cluster continually crashed on macOS. However, this setting might not apply to the single-node cluster. If that's the case, I'll move this note to the section for using Docker Compose.

Author: For a single-node cluster, I can run both the Elasticsearch container and the Kibana container with only 1.5 GB of memory, provided that no other containers are running. If I have another container running (like a local doc build) and try to start both Elasticsearch and Kibana, Docker stops one of those containers to regain memory.

you configure resource usage on the Advanced tab in Preference (macOS) or
Settings (Windows).
====

. Pull the {es} Docker image.
+
--
ifeval::["{release-state}"=="unreleased"]

WARNING: Version {version} of {es} has not yet been released, so no
Docker image is currently available for this version.

endif::[]

ifeval::["{release-state}"!="unreleased"]

[source,sh,subs="attributes"]
----
docker pull docker.elastic.co/elasticsearch/elasticsearch:{version}
----

endif::[]
--

. Pull the {kib} Docker image.
+
--
ifeval::["{release-state}"=="unreleased"]

WARNING: Version {version} of {kib} has not yet been released, so no
Docker image is currently available for this version.

endif::[]

ifeval::["{release-state}"!="unreleased"]

[source,sh,subs="attributes"]
----
docker pull docker.elastic.co/kibana/kibana:{version}
----

endif::[]
--

[[docker-start-enroll]]
[discrete]
==== Start a single-node {es} cluster and enroll {kib}

The following command starts a single-node {es} cluster for development or
testing.

. Start {es} in Docker. A password is generated for the `elastic` user and

> output to the terminal, plus an enrollment token for enrolling {kib}.

Reviewer 2: we generate a token for es nodes too

Author: ++ I'll change this to read:
>A password is generated for the `elastic` user and output to the terminal, plus enrollment tokens for enrolling {kib} and adding additional nodes to your cluster.

+
--
ifeval::["{release-state}"=="unreleased"]

WARNING: Version {version} of {es} has not yet been released, so no
Docker image is currently available for this version.

endif::[]

ifeval::["{release-state}"!="unreleased"]
[source,sh,subs="attributes"]
----

> docker run --name es-node01 --net elastic -p 9200:9200 -it docker.elastic.co/elasticsearch/elasticsearch:{version}

Reviewer 2: if we don't expose 9300, then no other node can join this cluster. I'm not sure we want to do this, but if we do we should point it out at least.

Reviewer: Actually, other nodes running on the same host can join the cluster. I'm not sure if we want to support multi-node from multiple hosts running docker in the getting started experience, as I assume that if you run with docker, you use the same host (but I may be wrong).

----

endif::[]
--
+
TIP: You might need to scroll back a bit in the terminal to view the password
and enrollment token.

. Copy the generated password and enrollment token and save them in a secure
location. These values are shown only when you start {es} for the first time.
+
[NOTE]
====
If you need to reset the password for the `elastic` user or other
built-in users, run the {ref}/reset-password.html[`elasticsearch-reset-password`]
tool. This tool is available in the {es} `bin` directory of the Docker container.

For example, to reset the password for the `elastic` user:

[source,sh]
----
docker exec -it es-node01 /usr/share/elasticsearch/bin/elasticsearch-reset-password -u elastic
----
====

. Copy the `http_ca.crt` security certificate from your Docker container to
your local machine. 
+
[source,sh]
----
docker cp es-node01:/usr/share/elasticsearch/config/tls_auto_config_<timestamp>/http_ca.crt .
----
`<timestamp>`:: The timestamp of when the auto-configuration process created
the security files directory in your Docker container. Use the Docker CLI to

> locate the timestamp for this directory.

Reviewer 2: We will see about changing this to be a standard dir name, but for now we can add the command for the user's reference:
```
docker exec -ti es-node01 /bin/bash -c "find /usr/share/elasticsearch -name http_ca.crt"
``

Author: Fancy! Thanks for this command; I'll definitely incorporate it into the workflow.

[[get-started-stack-docker]]
== Running the {stack} ("ELK") on Docker
++++
<titleabbrev>Running the {stack} on Docker</titleabbrev>
++++

The Elastic Docker registry contains Docker images for all the products in
the {stack}: https://www.docker.elastic.co/.

You can <<run-docker-secure,start the {stack} in Docker>> with security enabled
and configured by default. This option is great for quickly getting started with
{es} and {kib} for testing and development.

You can also <<get-started-docker-tls,start the {stack} with Docker Compose>> to
create a secured, multi-node cluster with a connected {kib} instance. This

> option requires more manual steps, but results in a more resilient cluster with

Reviewer 2: we're trying to make it a single step too, so maybe we don't need defensive language here anymore ?

Author: I'll update this language in the subsequent PR that introduces the simplified Docker Compose configuration.

Docker image is currently available for this version.

endif::[]

ifeval::["{release-state}"!="unreleased"]
[source,sh,subs="attributes"]
----
docker run --name es-node01 --net elastic -p 9200:9200 -it docker.elastic.co/elasticsearch/elasticsearch:{version}
----

endif::[]
--
+
TIP: You might need to scroll back a bit in the terminal to view the password
and enrollment token.

. Copy the generated password and enrollment token and save them in a secure
location. These values are shown only when you start {es} for the first time.

. Locate the CA certificate in your container. This command returns the path
to the `http_ca.crt` certificate, which you'll use in the next step.
+
[source,sh]
----
docker exec -it es-node01 /bin/bash -c "find /usr/share/elasticsearch -name http_ca.crt"
----

. Copy the `http_ca.crt` security certificate from your Docker container to
your local machine. 
+
[source,sh]
----
docker cp es-node01:/usr/share/elasticsearch/config/tls_auto_config_<timestamp>/http_ca.crt .
----
`<timestamp>`:: The timestamp of when the auto-configuration process created
the security files directory in your Docker container.

> 

Reviewer 2: we will update that as we remove the timestamp from the dirname, but I haven't had the time to do this change yet. I'll make sure I'll adjust docs as part of that change

Author: We'll need to make that change across the docs where it's used, but will be an easy change.

docker pull docker.elastic.co/kibana/kibana:{version}
----

endif::[]
--

[[docker-start-enroll]]
[discrete]
==== Start a single-node {es} cluster and enroll {kib}

The following command starts a single-node {es} cluster for development or
testing.

. Create an `elastic` network for your containers.
+
[source,sh]
----
docker network create elastic
----

. Start {es} in Docker. A password is generated for the `elastic` user and
output to the terminal, plus enrollment tokens for enrolling {kib} and adding
additional nodes to your cluster.
+
--
ifeval::["{release-state}"=="unreleased"]

WARNING: Version {version} of {es} has not yet been released, so no
Docker image is currently available for this version.

endif::[]

ifeval::["{release-state}"!="unreleased"]
[source,sh,subs="attributes"]
----
docker run --name es-node01 --net elastic -p 9200:9200 -it docker.elastic.co/elasticsearch/elasticsearch:{version}
----

endif::[]
--
+
TIP: You might need to scroll back a bit in the terminal to view the password
and enrollment token.

. Copy the generated password and enrollment token and save them in a secure
location. These values are shown only when you start {es} for the first time.
+

> [NOTE]

Reviewer 3: I think I'd pull this info about resetting the elastic password out of the procedure into a separate subsection.

Author: I added a new section for generating passwords and enrollment tokens and moved the relevant information there. I'll probably make this change in the Elasticsearch docs as well.

Preview link: https://stack-docs_1888.docs-preview.app.elstc.co/guide/en/elastic-stack-get-started/master/get-started-stack-docker.html#docker-generate
